8367142: Avoid InstanceKlass::cast when converting java mirror to InstanceKlass

iklam opened the pull request:
The purpose of this PR is to simplify JNI code and also to avoid unnecessary `InstanceKlass::cast()` calls by adding a new  function:

```
static InstanceKlass* java_lang_Class::as_InstanceKlass(oop java_class);
```

This PR is intended to be a strict clean-up that preserves existing behaviors.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8367142](https://bugs.openjdk.org/browse/JDK-8367142): Avoid InstanceKlass::cast when converting java mirror to InstanceKlass (**Enhancement** - P4)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/27158/head:pull/27158` \
`$ git checkout pull/27158`

Update a local copy of the PR: \
`$ git checkout pull/27158` \
`$ git pull https://git.openjdk.org/jdk.git pull/27158/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 27158`

View PR using the GUI difftool: \
`$ git pr show -t 27158`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/27158.diff">https://git.openjdk.org/jdk/pull/27158.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/27158#issuecomment-3268927680)
</details>

dholmes-ora reviewed (changes requested):
Sorry but I think this PR is trying to do too many things at once.

It is pushing JNI resolving inside internal JVM methods, which I think is a bad thing - we resolve JNI references at the boundary to get the oop that the VM wants to work with. Internal APIs should be oblivious to jobject and friends IMO. Also there may be times that the JNI/JVM method needs get the oop itself before extracting the  klass.

You are converting klass to instanceKlass where it has to be instanceKlass e.g. with redefinition. This is a good thing, but it is a very distinct thing that deserves its own cleanup (as per previous changes in that area).

You are defining a helper `java_lang_Class::as_InstanceKlass` to internalize the cast - this is fine but again a simple cleanup that would be better standalone IMO.

It is also not clear that JVM/JNI API's are properly checking that the incoming jobject is in fact a class of the right kind (ie not an array class object).

iklam commented:
> Sorry but I think this PR is trying to do too many things at once.
> 
> It is pushing JNI resolving inside internal JVM methods, which I think is a bad thing - we resolve JNI references at the boundary to get the oop that the VM wants to work with. Internal APIs should be oblivious to jobject and friends IMO. Also there may be times that the JNI/JVM method needs get the oop itself before extracting the klass.

There are 92 header files that have the word `jobject` in them. There are 71 cpp/hpp files with the word `JNIHandles::resolve` in them. So I am not sure if we really have that separation anymore.

There are 54 cases where we call `as_Klass(JNIHandles::resolve_non_null(x))`, so I think it would be nice to have a way to write less code.

> You are converting klass to instanceKlass where it has to be instanceKlass e.g. with redefinition. This is a good thing, but it is a very distinct thing that deserves its own cleanup (as per previous changes in that area).


I can move the considering_redefinition changes in a follow-up PR.

> You are defining a helper `java_lang_Class::as_InstanceKlass` to internalize the cast - this is fine but again a simple cleanup that would be better standalone IMO.
> 
> It is also not clear that JVM/JNI API's are properly checking that the incoming jobject is in fact a class of the right kind (ie not an array class object).

I am just converting

```
InstanceKlass* ik = InstanceKlass::cast(as_Klass(mirror));
```

to 

```
InstanceKlass* ik = as_InstanceKlass(mirror);
```

The code already assumes that it has an InstanceKlass, and I am not changing that.

iklam commented:
> > So I am not sure if we really have that separation anymore.
> 
> I think it is more that there are many bits of code that actually form the "boundary" (prims, services, some runtime, jvmci, interpreter-related). But I guess it is hard to argue this makes it markedly worse.

Arguably the translation of Java mirrors to Klasses is also a boundary (from Java representation to VM representation) :-)

In reality I think because jobjects are easy to use and are just another kind of handle (like Handle and OopHandle), the leakage from JNI code to other parts of VM just happened naturally.

> > The code already assumes that it has an InstanceKlass, and I am not changing that.
> 
> Okay.

BTW I removed the JVMTI changes from this PR.

dholmes-ora commented:
> Arguably the translation of Java mirrors to Klasses is also a boundary (from Java representation to VM representation) :-)

The mirror is an oop, both oop and kklass are internal VM representations.

stefank commented:
> There are 54 cases where we call as_Klass(JNIHandles::resolve_non_null(x)), so I think it would be nice to have a way to write less code.

I think you should deal with that by adding a helper function inside jni.cpp instead of extending the *Klass functions to take a jobject.

dholmes-ora commented on src/hotspot/share/prims/jvm.cpp:
```diff
@@ -840,14 +840,9 @@ JVM_ENTRY(jclass, JVM_FindClassFromClass(JNIEnv *env, const char *name,
-    Klass* to_class = java_lang_Class::as_Klass(mirror);
-    const char * to = to_class->external_name();
-    log_debug(class, resolve)("%s %s (verification)", from_name, to);
+    const char * from_name = java_lang_Class::as_Klass(from)->external_name();
+    const char * to_name = java_lang_Class::as_Klass(result)->external_name();
```
```suggestion
    const char* from_name = java_lang_Class::as_Klass(from)->external_name();
    const char* to_name = java_lang_Class::as_Klass(result)->external_name();
```
pre-existing nit

dholmes-ora commented on src/hotspot/share/prims/jvm.cpp:
```diff
@@ -911,12 +906,11 @@ static jclass jvm_lookup_define_class(jclass lookup, const char *name,
-  Klass* lookup_k = java_lang_Class::as_Klass(JNIHandles::resolve_non_null(lookup));
-  // Lookup class must be a non-null instance
+  InstanceKlass* lookup_k = java_lang_Class::as_InstanceKlass(lookup);
+  // Lookup class must not be a primitive class (whose mirror null Klass*)
```
```suggestion
  // Lookup class must not be a primitive class (whose mirror is a null Klass*)
```

dholmes-ora commented on src/hotspot/share/prims/jvm.cpp:
```diff
@@ -911,12 +906,11 @@ static jclass jvm_lookup_define_class(jclass lookup, const char *name,
+  // Lookup class must not be a primitive class (whose mirror null Klass*)
   if (lookup_k == nullptr) {
-    THROW_MSG_NULL(vmSymbols::java_lang_IllegalArgumentException(), "Lookup class is null");
+    THROW_MSG_NULL(vmSymbols::java_lang_IllegalArgumentException(), "Lookup class is primitive");
```
This is a behavioural change.

iklam replied:
I reverted the change to the error message.

I don't know how we will ever get a primitive class in there and who would be reading the error message. I added a comment saying the error message is wrong, so people reading this code will not get confused.

dholmes-ora commented on src/hotspot/share/prims/whitebox.cpp:
```diff
@@ -2155,20 +2155,19 @@ WB_ENTRY(jboolean, WB_IsSharedInternedString(JNIEnv* env, jobject wb, jobject st
-  Klass *k = java_lang_Class::as_Klass(JNIHandles::resolve_non_null(clazz));
-  if (!k->is_instance_klass()) {
-    return;
+  Klass *k = java_lang_Class::as_Klass(clazz);
```
```suggestion
  Klass* k = java_lang_Class::as_Klass(clazz);
```

dholmes-ora commented on src/hotspot/share/prims/whitebox.cpp:
```diff
@@ -2155,20 +2155,19 @@ WB_ENTRY(jboolean, WB_IsSharedInternedString(JNIEnv* env, jobject wb, jobject st
-    return;
+  Klass *k = java_lang_Class::as_Klass(clazz);
+  if (k->is_instance_klass()) {
+    InstanceKlass *ik = InstanceKlass::cast(k);
```
```suggestion
    InstanceKlass* ik = InstanceKlass::cast(k);
```

coleenp reviewed (changes requested):
jfr also is a boundary between Java -> Native code. That can have jobjects too, but resolve them before calling javaClasses.

iklam commented:
I didn't realize that my attempt to remove the `JNIHandles::resolve()` boilerplate can be conversional. I can't put a helper function in jni.cpp because this pattern is used in several files. I've reverted to the old code that makes the explicit calls to `JNIHandles::resolve()`.

I updated the JBS issue text as this PR is now only for reducing the number of InstanceKlass::cast() calls.

iklam commented:
Thanks @dholmes-ora @coleenp for the review
/integrate
